Source/DafnyLanguageServer/Handlers/DafnyHoverHandler.cs

```diff
@@ -24,30 +30,145 @@ public class DafnyHoverHandler : HoverHandlerBase {
+          var information = "";
+          foreach (var assertionBatch in node.AssertionBatches) {
+            assertionBatchIndex += 1;
+            if (!assertionBatch.Range.Contains(position)) {
+              continue;
+            }
+
+            var assertionIndex = 0;
+            var assertions = assertionBatch.Children.OfType<AssertionNodeDiagnostic>().ToList();
+            foreach (var assertionNode in assertions) {
+              if (assertionNode.Range.Contains(position) ||
+                  assertionNode.ImmediatelyRelatedRanges.Any(range => range.Contains(position))) {
+                var assertCmd = assertionNode.GetAssertion();
+                var batchRef = AddAssertionBatchDocumentation("batch");
+                var assertionBatchTime = assertionBatch.TimeSpent;
+                var assertionBatchResourceCount = assertionBatch.ResourceCount;
+                var assertionCount = assertionBatch.Children.Count;
+
+                var assertionId = assertionCount == 1 ? "" : $" #{assertionIndex + 1}/{assertionCount}";
+                var assertionInfo = $" of {batchRef} #{assertionBatchIndex + 1}/{assertionBatchCount} checked in {assertionBatchTime:n0}ms with {assertionBatchResourceCount:n0} resource count";
+
+                information += $"assertion{assertionId}{assertionInfo}:  \n";
+                // Not the main error displayed in diagnostics
+                if (!(assertionNode.GetCounterExample() is ReturnCounterexample returnCounterexample2 &&
+                      returnCounterexample2.FailingReturn.tok.GetLspRange().Contains(position))) {
+                  information += assertionNode.SecondaryPosition != null
+                    ? $"`{Path.GetFileName(assertionNode.Filename)}({assertionNode.SecondaryPosition.Line + 1}, {assertionNode.SecondaryPosition.Character + 1}): `"
+                    : "";
+                }
+
+                information += "*";
+                var statusMessage = assertionNode.StatusVerification switch {
+                  VerificationStatus.Verified => "Verified",
+                  VerificationStatus.Inconclusive => "Could not be checked",
+                  VerificationStatus.Error => assertCmd?.ErrorMessage ?? "Might not hold",
+                  _ => "Not checked",
+                };
+                var counterexample = assertionNode.GetCounterExample();
+
+                void SetDescription(ProofObligationDescription? description) {
+                  if (description == null) {
+                    return;
+                  }
+
+                  statusMessage = assertionNode?.StatusVerification switch {
+                    VerificationStatus.Verified => description.SuccessDescription,
+                    VerificationStatus.Error => description.FailureDescription,
+                    _ => statusMessage
+                  };
+                }
+
+                if (counterexample is ReturnCounterexample returnCounterexample) {
```

Comment:
This is where I hope I can properly use @atomb 's work to get more precise error messages